### PR TITLE
tests: ubuntu-image 2.0 compatibility fixes

### PR DESCRIPTION
**Pull request by bboozzoo**
We actually want EXTRA_FUNDAMENTAL to be word split as it may contain multiple
arguments, see a related failure when preparing core 20:

```
+ mv core20.snap /home/image/core20.snap
+ EXTRA_FUNDAMENTAL='--extra-snaps /home/image/pc-kernel_*.snap --snap /home/image/core20.snap'
+ /snap/bin/ubuntu-image snap -w /home/image /home/image/pc.model --channel edge '--extra-snaps /home/image/pc-kernel_*.snap --snap /home/image/core20.snap' --extra-snaps /home/image/snapd_2.53.2+git628.g35bb479-dirty_amd64.snap --output /home/image/core20-amd64.img
Error: unknown flag `extra-snaps /home/image/pc-kernel_*.snap --snap /home/image/core20.snap'
-----
```

### CODE REVIEW

**Review by mvo5 — APPROVED** [review]
Thank you

**Review by stolowski — APPROVED** [review]
LGTM

**Review by sergiocazzolato — APPROVED** [review]
tx

**Review by anonymouse64 — APPROVED** [review]
lgtm, but AIUI we also still want ubuntu-image 1.11 because the new go based one is not working with uc18 at least for some reason

**Review by mardy — DISMISSED** [review]
LGTM!

**Comment by anonymouse64** [comment]
seems some tests are still unhappy, this PR is still using candidate which is 2.0 AIUI

**Comment by bboozzoo** [comment]
> seems some tests are still unhappy, this PR is still using candidate which is 2.0 AIUI

Yes, that's intended, a fix using latest/stable landed in master. This branch reverts back to candidate and add bits for compatibility

**Comment by anonymouse64** [comment]
We also need to run nested tests with the new ubuntu-image, some of these will fail as discussed, since ubuntu-image will not be built with the fakestore test keys built into it. The suggestion for now was just to rebuild the ubuntu-image snap for these tests in order to inject the keys.

**Comment by bboozzoo** [comment]
```
+ /snap/bin/ubuntu-image snap --image-size 10G /home/gopath/src/github.com/snapcore/snapd/tests/lib/assertions/developer1-20-secured.model --output-dir /tmp/work-dir/images --snap /tmp/work-dir/assets/pc-kernel_5.4.0-92.103.1_amd64.snap --snap /home/gopath/src/github.com/snapcore/snapd/tests/nested/manual/core20-grade-signed-above-testkeys-boot/snapd-from-deb.snap --snap /home/gopath/src/github.com/snapcore/snapd/tests/nested/manual/core20-grade-signed-above-testkeys-boot/new-core20.snap --snap /home/gopath/src/github.com/snapcore/snapd/tests/nested/manual/core20-grade-signed-above-testkeys-boot/extra-snaps/pc_20-0.4_amd64.snap
Error: Error preparing image: cannot override channels, add devmode snaps, local snaps, or extra snaps with a model of grade higher than dangerous
```
this one looks like it's intentional

**Comment by bboozzoo** [comment]
And someone in the forum had a similar problem? https://forum.snapcraft.io/t/cannot-build-image-error-preparing-image-cannot-override-channels-add-devmode-snaps-local-snaps-or-extra-snaps-with-a-model-of-grade-higher-than-dangerous/27836

**Comment by bboozzoo** [comment]
FWIW I suspect the problem with ubuntu-image is because it was not built with testkeys, thus assertions signed by testrootorg/developer1 would be rejected.

**Comment by bboozzoo** [comment]
I've pushed patches that build ubuntu-image from Go source with `withtestkeys` tags, thus skipping the ubuntu-image snap completely, but it should make it possible to build signed/secured model images again. This probably warrants an agreement whether that's the right way to go.

File: spread.yaml [hunk 1]
```diff
@@ -922,8 +919,14 @@ suites:
...
             . "$TESTSLIB"/pkgdb.sh
             distro_update_package_db
-            distro_install_package snapd qemu qemu-utils genisoimage sshpass qemu-kvm cloud-image-utils ovmf kpartx xz-utils
-            snap install --classic --channel="$UBUNTU_IMAGE_SNAP_CHANNEL" ubuntu-image
+            distro_install_package snapd qemu qemu-utils genisoimage sshpass qemu-kvm cloud-image-utils ovmf kpartx xz-utils mtools
+            # build a custom version ubuntu-image with test keys
+            (
+                export GO111MODULE=off
+                # use go get so that ubuntu-image is built with current snapd sources
+                go get github.com/canonical/ubuntu-image/cmd/ubuntu-image
```
Comment (anonymouse64) [inline spread.yaml:927]: it would be nice to fetch a specific tag/branch/etc rather than just pulling straight from main, but not sure what is appropriate

File: spread.yaml [hunk 2]
```diff
@@ -37,10 +37,7 @@ environment:
     MANAGED_DEVICE: "false"
     # a global setting for LXD channel to use in the tests
     LXD_SNAP_CHANNEL: "latest/edge"
-    # TODO: as of 2021-11-18 ubuntu-image in candidate and up has a bug which is
-    # fixed in https://github.com/canonical/ubuntu-image/pull/26, revert to
-    # latest/candidate when the fix lands in master and the snap is released
-    UBUNTU_IMAGE_SNAP_CHANNEL: "latest/stable"
+    UBUNTU_IMAGE_SNAP_CHANNEL: "latest/edge"
```
Comment (anonymouse64) [inline spread.yaml:40]: do we even need this env var anymore?

**Review by anonymouse64 — COMMENTED** [review]
lgtm, only concern I have is following main git branch which is probably as close as we can get to following the "bleeding edge" of ubuntu-image, but I think it's probably fine for now and we can think about a better risk management strategy for ubuntu-image later on (maybe we could download/install the snap from beta/candidate and check what git commit that was built with and use that as our base?)

**Review by anonymouse64 — APPROVED** [review]

File: spread.yaml (hunk 1 above)
Comment (bboozzoo) [reply spread.yaml:927]: There are tags which correspond to ubuntu-image package versions, eg. `2.1+22.04ubuntu2`

**Comment by bboozzoo** [comment]
Turns out we cannot universally use ubuntu-image built directly from Go, as it runs `mkfs.ext4 -d` when creating the image. However, the version of e2fsprogs shipped with 16.04 do not did not support `-d` :cry: Fortunately, AFAICT we do not need to run the tests that would fail if ubuntu-image does not know about test keys on 16.04, so I'm hoping this next run will be successful and we can finally merge this branch.

File: spread.yaml (hunk 2 above)
Comment (bboozzoo) [reply spread.yaml:40]: actually it's needed again :/ (at least for 16.04)

Comment (anonymouse64) [reply spread.yaml:40]: ack makes sense

File: tests/core/swapfiles/task.yaml [hunk 3]
```diff
@@ -5,10 +5,11 @@ execute: |
   snap set system swap.size=200M
 
   echo "Check that the swap file was setup"
-  retry -n 60 --wait 1 bash -c "cat /proc/swaps | MATCH '\s+file\s+204796'"
+  # use swapon as it rounds up the numbers nicely
+  retry -n 60 --wait 1 bash -c "swapon --show | MATCH '\s+file\s+200M\s'"
```
Comment (anonymouse64) [inline tests/core/swapfiles/task.yaml:9]: is this related to ubuntu-image ?

**Review by anonymouse64 — APPROVED** [review]
lgtm, wondering why the swap changes?

File: tests/core/swapfiles/task.yaml (hunk 3 above)
Comment (bboozzoo) [reply tests/core/swapfiles/task.yaml:9]: Yes, for some reason there were byte size changes here (204796 -> 204784, 3 pages?). I suspect this can be caused by a different fs configuration, as looking at /usr/bin/mkswapfile the could should do the right and create 200MB sized file using fallocate and then runs mkswap on that. OTOH swapon --show correctly reports 200M size in both cases.

File: tests/lib/prepare.sh [hunk 4]
```diff
@@ -843,12 +843,25 @@ setup_reflash_magic() {
     UNPACK_DIR="/tmp/$core_name-snap"
     unsquashfs -no-progress -d "$UNPACK_DIR" /var/lib/snapd/snaps/${core_name}_*.snap
 
-    # install ubuntu-image
-    snap install --classic --channel="$UBUNTU_IMAGE_SNAP_CHANNEL" ubuntu-image
+    if os.query is-core16; then
+        # the new ubuntu-image expects mkfs to support -d option, which was not
+        # supported yet by the version of mkfs that shipped with Ubuntu 16.04
+        snap install ubuntu-image --channel="$UBUNTU_IMAGE_SNAP_CHANNEL" --classic
```
Comment (mardy) [inline tests/lib/prepare.sh:849]: maybe we should define (and export) the `UBUNTU_IMAGE` variable at this point, since we currently define it in two places?

**Review by mardy — APPROVED** [review]
LGTM, added a minor question.

File: tests/core/swapfiles/task.yaml (hunk 3 above)
Comment (anonymouse64) [reply tests/core/swapfiles/task.yaml:9]: hmm interesting, but sure makes sense

File: tests/lib/prepare.sh (hunk 4 above)
Comment (bboozzoo) [reply tests/lib/prepare.sh:849]: It would have to be a separate helper in one of the scripts, just exporting it here would not make it availble in other shell fragments. OTOH maybe we could pull a trick and symlink to the right one, but that sounds like a followup material anyway.
